kong/tools/grpc.lua

```diff
@@ -1,152 +1,114 @@
-  end,
-  function (t)
-    if type(t) ~= "string" then
-      error(string_format(
-        "expected time string, got (%s)%q", type(t), tostring(t)))
-    end
-
-    local ds = date(t) - epoch
-    return {
-      seconds = ds:spanseconds(),
-      nanos = ds:getticks() * 1000,
-    }
-  end)
-end
-
-function _M.new()
-  local protoc_instance = protoc.new()
-  -- order by priority
-  for _, v in ipairs {
-    "/usr/local/kong/include",
-    "/usr/local/opt/protobuf/include/", -- homebrew
-    "/usr/include",
-    "kong/include",
-    "spec/fixtures/grpc",
-  } do
-    protoc_instance:addpath(v)
   end
-  protoc_instance.include_imports = true
-
-  return setmetatable({
-    protoc_instance = protoc_instance,
-  }, _MT)
 end
 
-function _M:addpath(path)
-  local protoc_instance = self.protoc_instance
-  if type(path) == "table" then
-    for _, v in ipairs(path) do
-      protoc_instance:addpath(v)
+--- Executes function for every field in file
+-- Handles also nested types
+-- @param file File descriptor
+-- @param f Functiont to execute
+local function for_each_field( file, f )
+  local function parse_message( msg )
+    for _, field in pairs( msg.field or {} ) do
+      f( file, msg, field )
     end
 
-  else
-    protoc_instance:addpath(path)
+    -- traverse nested types
+    for _, nm in pairs( msg.nested_type or {} ) do
+      nm.full_name = msg.full_name .. "." .. nm.name
+      parse_message( nm )
+    end
   end
-end
 
-function _M:get_proto_file(name)
-  for _, path in ipairs(self.protoc_instance.paths) do
-    local fn = path ~= "" and path .. "/" .. name or name
-    local fh, _ = io.open(fn)
-    if fh then
-      return fh
-    end
+  for _, msg in pairs( file.message_type or {} ) do
+    msg.full_name = "." .. file.package .. "." .. msg.name
+    parse_message( msg )
   end
-  return nil
 end
 
-local function each_method_recur(protoc_instance, fname, f, recurse)
-  local parsed = protoc_instance:parsefile(fname)
-  if f then
-    if recurse and parsed.dependency then
-      if parsed.public_dependency then
-        for _, dependency_index in ipairs(parsed.public_dependency) do
-          local sub = parsed.dependency[dependency_index + 1]
-          each_method_recur(protoc_instance, sub, f, true)
-        end
-      end
-    end
+--- Add path to search for parsed proto files
+-- @param path Path to add
+function _M:add_path( path ) 
+  self.protoc:addpath( path )
+end
 
-    for _, srvc in ipairs(parsed.service or {}) do
-      for _, mthd in ipairs(srvc.method or {}) do
-        f(parsed, srvc, mthd)
-      end
+--- Traverse loaded file and call hooks if possible
+-- File is processed recursively according to `protoc.include_import` settings
+-- Hooks are optional
+-- @param filename Filename to parse
+-- @param method_hook Function to be called for each method
+-- @param field_hook Function to be called for each field (also in nested messages)
+function _M:parse_file( filename, method_hook, field_hook )
```

Comment:
This function should be more like an internal function, used to implement more simple functions. Thus this function should have a name like `traversal_proto`, and we use this function to implement `parse_file`, which does not require any callbacks.

Reply:
What about renaming it to `traverse_proto_file` (to be consistent with `open_proto_file` later)?

Reply:
https://github.com/Kong/kong/pull/8297/commits/c921f9a455824a326589bb3dbd44cbdaf81aa366

Reply:
I think that should do. It's more of a problem with the original code.